typescript/src/redemption.ts

```diff
@@ -406,3 +410,103 @@ export async function getRedemptionRequest(
 
   return redemptionRequests[0]
 }
+
+/**
+ * Finds the oldest active wallet that has enough BTC to handle a redemption request.
+ * @param amount The amount to be redeemed in satoshis.
+ * @param bridge The handle to the Bridge on-chain contract.
+ * @param bitcoinClient Bitcoin client used to interact with the network.
+ * @param bitcoinNetwork Bitcoin network.
+ * @returns Promise with the wallet details needed to request a redemption.
+ */
+export async function findWalletForRedemption(
+  amount: BigNumber,
+  bridge: Bridge,
+  bitcoinClient: BitcoinClient,
+  bitcoinNetwork: BitcoinNetwork
+): Promise<{
+  walletPublicKey: string
+  mainUtxo: UnspentTransactionOutput
+}> {
+  const wallets = await bridge.getNewWalletRegisteredEvents()
+
+  let walletData:
+    | {
+        walletPublicKey: string
+        mainUtxo: UnspentTransactionOutput
+      }
+    | undefined = undefined
+  let maxAmount = BigNumber.from(0)
+
+  for (const wallet of wallets) {
+    const { state, mainUtxoHash, walletPublicKey } = await bridge.wallets(
+      wallet.walletPublicKeyHash
+    )
+
+    // Wallet must be in Live state.
+    if (
+      state !== WalletState.Live ||
+      mainUtxoHash.equals(
+        Hex.from(
+          "0x0000000000000000000000000000000000000000000000000000000000000000"
+        )
+      )
+    ) {
+      console.debug(
+        `Wallet is not in Live state or main utxo is not set. \
+        Continue the loop execution to the next wallet...`
+      )
+      continue
+    }
+
+    const walletBitcoinAddress = encodeToBitcoinAddress(
+      wallet.walletPublicKeyHash.toString(),
+      true,
+      bitcoinNetwork
+    )
+
+    const utxos = await bitcoinClient.findAllUnspentTransactionOutputs(
+      walletBitcoinAddress
+    )
+
+    // We need to find correct utxo- utxo components must point to the recent
+    // main UTXO of the given wallet, as currently known on the chain.
+    const mainUtxo = utxos.find((utxo) =>
+      mainUtxoHash.equals(bridge.buildUtxoHash(utxo))
+    )
+
+    if (!mainUtxo) {
+      console.debug(
+        `Could not find matching UTXO on chains for wallet public key hash \
+        (${wallet.walletPublicKeyHash.toPrefixedString()}). Continue the loop \
+        execution to the next wallet...`
+      )
+      continue
+    }
+
+    // Save the max possible redemption amount.
+    maxAmount = mainUtxo.value.gt(maxAmount) ? mainUtxo.value : maxAmount
+
+    if (mainUtxo.value.gte(amount)) {
+      walletData = {
+        walletPublicKey: walletPublicKey.toString(),
+        mainUtxo,
+      }
+
+      break
+    }
+
+    console.debug(
+      `The wallet (${wallet.walletPublicKeyHash.toPrefixedString()}) \
```

Comment:
Formatting has to be fixed, as this message was logged like this:
```
The wallet (0x328d992e5f5b71de51a1b40fcc4056b99a88a647)       cannot handle the redemption request. Continue the loop execution to the       next wallet...
```

Reply:
0f11838